src/wallet/rpc/addresses.cpp

```diff
@@ -66,6 +70,137 @@ RPCHelpMan getnewaddress()
+{
+    return RPCHelpMan{"getsilentaddress",
+                "\nReturns a silent paymet address if it is enabled in wallet.\n"
+                "If 'label' is specified and it is new, a new address will be generated. \n"
+                "If the 'label' is already assigned to a silent address, this same address will be shown.\n",
+                {
+                    {"label", RPCArg::Type::STR, RPCArg::Default{""}, "The label name for the address to be linked to. It can also be set to the empty string \"\" to represent the default label. The label does not need to exist, it will be created if there is no label by the given name."},
+                },
+                RPCResult{
+                    RPCResult::Type::OBJ, "", "",
+                    {
+                        {RPCResult::Type::STR, "address", "The silent address"},
+                        {RPCResult::Type::STR_HEX, "label", "The label assigned to this silent address"},
+                        {RPCResult::Type::NUM, "identifier", "The identifier corresponding to the given label"},
+                        {RPCResult::Type::STR, "warnings", /*optional=*/true, "Any warnings associated with the address"},
+                    }
+                },
+                RPCExamples{
+                    HelpExampleCli("getsilentaddress", "")
+            + HelpExampleRpc("getsilentaddress", "")
+                },
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
+    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
+    if (!pwallet) return UniValue::VNULL;
+
+    LOCK(pwallet->cs_wallet);
+
+    if (!pwallet->IsWalletFlagSet(WALLET_FLAG_SILENT_PAYMENT)) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Error: This wallet is not enabled for silent payments");
+    }
+
+    if (!pwallet->CanGetAddresses()) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Error: This wallet has no available keys");
+    }
+
+    // Parse the label first so we don't generate a key if there's an error
+    std::string label;
+    if (!request.params[0].isNull())
+        label = LabelFromValue(request.params[0]);
+
+    UniValue ret(UniValue::VOBJ);
+    for (auto const& [key, val] : pwallet->m_silent_address_book)
+    {
+        if (val.m_label == label) {
+            ret.pushKV("address", val.m_address);
+            ret.pushKV("label", val.m_label);
+            ret.pushKV("identifier", key);
+            return ret;
+        }
+    }
+
+    const auto result = pwallet->GetSilentDestination(label);
+
+    if (!result) {
+        throw JSONRPCError(RPC_WALLET_ERROR, util::ErrorString(result).original);
+    }
+
+    const auto& [address, identifier] = *result;
+
+    ret.pushKV("address", address);
+    ret.pushKV("label", label);
+    ret.pushKV("identifier", identifier);
+    if (identifier > 0) {
+        ret.pushKV("warnings", "This address is not a new identity. It is a re-use of an existing identity with a different label.");
```

Comment:
The term "identity" is confusing here. It's also confusing that you get this error when you generate a new silent payment address, but not when you request an existing one (by reusing a label).

I guess what you're trying to say here is that an outside observer can tell that two silent payment addresses belong to the same person? Maybe just say that:

`The silent payment addresses are unique for accounting purposes only. They can be linked to the same wallet. For privacy a new wallet is required.`

That said, I think this feature is a footgun.